bsip-0040.md

```diff
@@ -0,0 +1,379 @@
+for each operation if there is a `custom active authority` for any of its required accounts. Check for every required 
+account of the transaction if all its belonging operations have at least one positively matched `custom active authority` 
+(see Specifications for definition of "matching"), and if so behave as if the active authority of the corresponding account is present (recursive authorities are excluded, see Examples).
+
+# Specification
+
+All descriptions in this section are on a pseudo/prosa level and no recommendation how it can best be implemented or serialized. They are meant to facilitate the understanding. If anything in the looping process or order of evaluation is unsuitable for actual implementation, changes can be made accordingly as long the same functionality is achieved.
+
+### Custom active permission and custom active authority
+
+A `custom active permission` contains a list of `custom active authorities` and looks like follows (in JSON-like/pseudo for clarification):
+```
+custom_active_permission = {
+    account_id,    // account that is assigned to this permission
+    authorities = list of custom_active_authority objects
+}
+custom_active_authority = {
+    enabled,       // status of this authority, true or false. true when created
+    valid_from,    // timestamp when this is active
+    valid_to,      // timestamp when this is invalid
+    operation_id,   // operation id of the target operation,
+    authority,     // same as for the existing authotities (multisig with weighted accounts or keys),
+    restrictions        // see below
+}
+```
+Note: This assumes `custom_active_permission` is stored in a separate index. Actual implementation details left to the implementer, as long as every `custom_active_permission` can be assigned to exactly one account.
+
+A `custom active authority` is matched against operations of an incoming, signed transaction. The wording *matching* refers to:
+- `operation_id` is equal to the id of the incoming operation
+- `account_id` is in the required accounts of the operation
+- the `authority` of the `custom_active_authority` is given by the signatures of the transaction
+- timestamp `now` is within `valid_to` and `valid_from` 
+- all `restrictions` assert positively
+
+### Restrictions
+
+The `restrictions` field is a list of restrictions consisting of argument to assert mappings.
+A dictionary-type object like
+```
+restriction = {
+   function,     // argument_identifier
+   argument,     // pointer to a dynamic value (argument of the operation, or attribute when nested)
+   data,         // data specific to the function
+}
+```
+is called a restriction, that can assert itself positively (passed) or negatively (violated). All restrictions are evaluated per default with `AND` logic to determine if the whole list has asserted positively. 
+
+List of possible restrictions are:
```

Comment:
I think it's better to expand this list a bit.
* Good to be able to use a function on the member, E.G. "length of" a string or a list, "size of" an object, and etc.
* There are `optional` fields in operations, which could have a special `is_valid` function, and perhaps a `is_not_valid` function. E.G. a custom authority can not use memo when transferring.
* There are vectors / lists, which don't fit `attribute_asset`, but could have `contains` / `does_not_contain` function .
* Good to have `equal` and `not_equal` functions to make restrictions simpler.

Examples:

| member | function | operator | argument | description |
| ------------- | ------------- | ----- | ---| --- |
| to | self | eq | "1.2.3" | "to" should only be "1.2.3" |
| memo | size | le | 100 | total memo size should be less than 100 bytes |
| active | self | not_valid | | "active" field should be not valid (when updating an account) |
| new_feed_producers | length | eq | 7 | feed producer list should have 7 items |
| new_feed_producers | self | contains | "1.2.4" | must have "1.2.4" as a feed producer |

Advanced expressions, E.G. for `limit_order_create_operation`:
```
{
      asset           fee;
      account_id_type seller;
      asset           amount_to_sell;
      asset           min_to_receive;
}
```
users may want to restrict `min_to_receive / amount_to_sell > X`, how to define it?

Reply:
> - Good to be able to use a modifier on the member, E.G. "length of" a string or a list, "size of" an object, and etc.

I have added `object` and `list` to the implicit type conversions of `lt` et al. Also I added `eq` as operator that also does the implicit number conversion.

> - There are optional fields in operations, which could have a special is_valid function, and perhaps a is_not_valid function. E.G. a custom authority can not use memo when transferring.

Can you elaborate what they are checking? To check if a value is NOT set we can use `any` with empty list value. It will pass if not given and fail if any value is given.

> - There are vectors / lists, which don't fit attribute_asset, but could have contains / does_not_contain function .

Added

> - Good to have equal and not_equal functions to make restrictions simpler.

I don't think that is necessary, use `any` and `none` with one-item list.

Reply:
TBH I don't like implicit conversions which will increase complexity and are error prone. 

For example, a `memo` is
```
{
  public_key    from,
  public_key    to,
  int           nonce,
  vector<char>  message
}
```
and in transfers we actually use `optional<memo>`. It doesn't make sense to specially convert it to a number for comparing to another number. It doesn't make sense as well to use "any" or "none" since it's a struct but not a simple data type so unable to specify any value in the list. "`any` function with an empty list value" is hard to understand, IMHO it means "this field should be any of nothing" thus should always fail.

Since we have `price_per_kbyte` in fee schedule for many operations, it makes sense to have a `pack_size` function to calculate size of a member attribute after serialized. `pack_size` and `size` and `length` are different metrics. IMHO, better explicitly give the tools to users rather than converting implicitly.

Reply:
More thoughts:

for `optional` type, we can define a special value `null` (which is also used in jsonification) for users to use.

* ` A eq null` means A can only be null
* `A any [ null, 1 ]` means A can be null or 1
* `A ge null` is invalid
* `A gt 1` means A must be a number, and is valid (not null), and greater than 1

Actually this is still too complex, it would require much efforts to implement an array type which can contain a `null`.

Reply:
> "any function with an empty list value" is hard to understand, IMHO it means "this field should be any of nothing" thus should always fail.

To be precise, it means "if set, this field should be any of nothing".

I will think about it and adjust the text. I will probably aim to add flexibility in the meaning of "this list of restrictions may be redefined to improve clarity,performance or to reduce complexity of implementation while maintaining the intended functionality".